codelists/tests/test_hierarchy.py

```diff
@@ -0,0 +1,386 @@
+        ("b", "d"),
+        ("b", "e"),
+        ("c", "e"),
+        ("c", "f"),
+    ]
+
+    return Hierarchy("a", edges)
+
+
+def build_hierarchy():
+    r"""Return hierarchy with this structure:
+
+           a
+          / \
+         b   c
+        / \ / \
+       d   e   f
+      / \ / \ / \
+     g   h   i   j
+    """
+
+    edges = [
+        ("a", "b"),
+        ("a", "c"),
+        ("b", "d"),
+        ("b", "e"),
+        ("c", "e"),
+        ("c", "f"),
+        ("d", "g"),
+        ("d", "h"),
+        ("e", "h"),
+        ("e", "i"),
+        ("f", "i"),
+        ("f", "j"),
+    ]
+
+    return Hierarchy("a", edges)
+
+
+def test_nodes():
+    hierarchy = build_small_hierarchy()
+
+    assert hierarchy.nodes == {"a", "b", "c", "d", "e", "f"}
+
+
+def test_parent_map():
+    hierarchy = build_small_hierarchy()
+
+    for node, children in {
+        "a": set(),
+        "b": {"a"},
+        "c": {"a"},
+        "d": {"b"},
+        "e": {"b", "c"},
+        "f": {"c"},
+    }.items():
+        assert hierarchy.parent_map[node] == children
+
+
+def test_child_map():
+    hierarchy = build_small_hierarchy()
+
+    for node, parents in {
+        "a": {"b", "c"},
+        "b": {"d", "e"},
+        "c": {"e", "f"},
+        "d": set(),
+        "e": set(),
+        "f": set(),
+    }.items():
+        assert hierarchy.child_map[node] == parents
+
+
+def test_ancestors():
+    hierarchy = build_small_hierarchy()
+
+    for node, ancestors in {
+        "a": set(),
+        "b": {"a"},
+        "c": {"a"},
+        "d": {"a", "b"},
+        "e": {"a", "b", "c"},
+        "f": {"a", "c"},
+    }.items():
+        assert hierarchy.ancestors(node) == ancestors
+
+
+def test_descendants():
+    hierarchy = build_small_hierarchy()
+
+    for node, descendants in {
+        "a": {"b", "c", "d", "e", "f"},
+        "b": {"d", "e"},
+        "c": {"e", "f"},
+        "d": set(),
+        "e": set(),
+        "f": set(),
+    }.items():
+        assert hierarchy.descendants(node) == descendants
+
+
+def test_walk_depth_first_as_tree_from_root():
+    hierarchy = build_small_hierarchy()
+
+    assert list(hierarchy.walk_depth_first_as_tree()) == [
+        ("a", 0, 0, 0, 1),
+        ("b", 1, 0, 1, 1),
+        ("d", 2, 0, 1, 1),
+        ("d", 2, 0, 1, -1),
+        ("e", 2, 1, 0, 1),
+        ("e", 2, 1, 0, -1),
+        ("b", 1, 0, 1, -1),
+        ("c", 1, 1, 0, 1),
+        ("e", 2, 0, 1, 1),
+        ("e", 2, 0, 1, -1),
+        ("f", 2, 1, 0, 1),
+        ("f", 2, 1, 0, -1),
+        ("c", 1, 1, 0, -1),
+        ("a", 0, 0, 0, -1),
+    ]
+
+    assert list(hierarchy.walk_depth_first_as_tree(starting_node="b")) == [
+        ("b", 0, 0, 0, 1),
+        ("d", 1, 0, 1, 1),
+        ("d", 1, 0, 1, -1),
+        ("e", 1, 1, 0, 1),
+        ("e", 1, 1, 0, -1),
+        ("b", 0, 0, 0, -1),
+    ]
+
+    assert list(
+        hierarchy.walk_depth_first_as_tree(
+            starting_node="b", sort_key=lambda label: -ord(label)
```

Comment:
What does `-ord(label)` do here?

Reply:
It sorts the labels in reverse alphabetic order.  Does it need a comment?

Reply:
Yes, please.  I've read the docs on `ord` at least 3 times trying to get my head around this.  I think I got there in the end but your comment succinctly captured it in a way that I'd not arrived at so probably for the best!

Reply:
Addressed in cb39649fbd4fafcb72910aef717c7339e1944a0f.